[Review] Add FREERTOSLWIP build to travis ci.

cmbahadir opened the pull request:
Created this PR with reference to [#2993 Comment](https://github.com/open62541/open62541/pull/2993#issuecomment-523176846) but with below concerns;

Still need to externally modify the LWIP_IPV6 macro with sed. See #2511 and #1910

tcpip.h compile error is ignored which does not have any effect on freertos build. It is raised by the used xtensa-gcc toolchain.

python version 2.7.5 raises warnings on line;
python -m pip install --user -r $IDF_PATH/requirements.txt

where there is no warning for 2.7.12 in my local here.
https://github.com/cmbahadir/opcua-esp32/blob/open62541_ci/.travis.yml

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open62541/open62541?pullRequest=3023) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open62541/open62541?pullRequest=3023) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open62541/open62541?pullRequest=3023) it.</sub>

jpfr commented:
Very nice!

This depends on your repository https://github.com/cmbahadir/opcua-esp32.git
So the build on our end depends on the new dependency being always up to date.

How much effort would it be to make the FreeRTOS build self-contained in our repo?
In the first step, it could be enough to build the library as shared objects and not make a "flashable" binary.

In a second step, we could use one of the FreeRTOS environments for Linux/POSIX to also run the unit tests for a FreeRTOS-y environment.

https://www.freertos.org/FreeRTOS-simulator-for-Linux.html
https://github.com/nqd/freertos_linux_devl

cmbahadir commented:
Hi @jpfr yes it is not good to be my repo dependent but was the fastest way. And probably ı wont update it for a while :)

I got to check how much work needed to make it self contained. I will let you know.

cmbahadir commented:
Hi @jpfr ,

I have checked the freertos posix simulator on linux and seen that it is not only enough to build the Freertos as shared object but also lwip needed and porting of lwip is specific to an architecture.

@cabralfortiss  
Linux with FREERTOS simulator results with  lwip/tcpip.h undefined reference error which also exists whenever the UA_ARCHITECTURE is selected as "FREERTOSLWIP" w/o any simulator.

I couldn't dive in the details but seems to me that the main bottleneck for a CI of FREERTOSLWIP is LWIP porting.

Pro commented on tools/travis/travis_linux_script.sh:
```diff
@@ -1,6 +1,21 @@
+    export IDF_PATH=/home/travis/build/open62541/esp-idf
+    cd $IDF_PATH/examples/opcua-esp32
+    cp /home/travis/build/open62541/open62541/build/open62541.c components/open62541lib
+    sed -i '/#define UA_IPV6 LWIP_IPV6/d' /home/travis/build/open62541/open62541/build/open62541.h
```
Instead of using sed to change the `LWIP_IPV6` define, can you modify your PR to change this line:
https://github.com/open62541/open62541/blob/46270f270d7a231e7be24711a760d271c77c5842/arch/common/ua_lwip.h#L30
into something like this:
```
#ifdef LWIP_IPV6
#define UA_IPV6 LWIP_IPV6
#endif

cmbahadir replied:
Macro added with https://github.com/open62541/open62541/pull/3023/commits/5f31676c03d8fe957e07f31f0cd336b460b4cec9

But needs review if its the right place to add.

Pro commented on tools/travis/travis_linux_before_install.sh:
```diff
@@ -1,6 +1,35 @@
+if [ "$ARCH" = "freertoslwip" ]; then
+    echo -e "\r\n==Compile multithreaded version==" && echo -en 'travis_fold:start:before_install.build.multithread\\r'
+    mkdir -p build && cd build
+    cmake \
```
cmake and make commands should not be executed in the install script, but in the `travis_linux.sh` script.

Then you should be able to build the whole open62541 stack and not ignore the build error

cmbahadir replied:
Fixed with https://github.com/open62541/open62541/pull/3023/commits/5f31676c03d8fe957e07f31f0cd336b460b4cec9

Pro commented on tools/travis/travis_linux_before_install.sh:
```diff
@@ -1,6 +1,35 @@
+    -DUA_ENABLE_PUBSUB:BOOL=ON \
+    -DUA_LOGLEVEL:STRING=600 ..
+	# Compile error ignored and does not cause any problems. Related Issues: #2887 and #2893
+    make -j || true
```
hmm, this does not look so good. If make fails, then it's bad. Shouldn't be ignored.

First installing esp-idf and then building the stack should solve it?

cmbahadir replied:
I couldn't find a way to compile open62541 library without this error and only needed files are amalgamated open62541.c/h files to build a flashable binary with xtensa toolchain, so;

```
/home/travis/build/cmbahadir/open62541/build/open62541.h:239:24: fatal error: lwip/tcpip.h: No such file or directory
 #include <lwip/tcpip.h>
```
error is ignored. 

There are discussions on #2887 and #2893.

cabralfortiss replied:
The problem is that he needs only the amalgamated files to then compile with the right tools. 
I see two possible solutions: 
1- The cleanest is to avoid this error by compiling open62541 properly as a library and then use the library in the project compilation (this instead of the using the amalgamated files in the project). This requires that open62541 is compiled with the xtensa tools, giving the right flags. Unfortunately I couldn't understand the "make" for the project later on. 
2- Create a CMake target in open62541 which only generates the amalgamated files and doesn't compile.

I see point 2 very specific. Point 1 is the best option, but also requires a good understanding of the building process of each part of the system.

cmbahadir replied:
I would suggest to go with option 2 since build systems for each target maybe different. So rather than covering every toolchains configurations with different flags, it would be  enough to provide the the get amalgamated .c./.h files. 

And it will make the current "tcpip.h" error invisible to the user., so user will be more confident about what he is doing.

Just an opinion. :)

Pro commented on tools/travis/travis_linux_before_install.sh:
```diff
@@ -1,6 +1,35 @@
+	wget https://dl.espressif.com/dl/xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz
+	tar -xzf xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz
+	cd .. 
+	git clone https://github.com/cmbahadir/opcua-esp32.git opcua-esp32
```
Why do we need this additional dependency/repo?

cmbahadir replied:
> 4. Optional step in a second PR when the previous ones are implemented: use the xtensa toolchain with an example code to build a flashable binary

Actually this repo of me is an example code to build a flashable binary.

cmbahadir replied:
Its a sample application to build a flashable binary with xtensa-toolchain.

Pro commented on tools/travis/travis_linux_before_install.sh:
```diff
@@ -11,7 +40,7 @@ if ! [ -z ${CLANG_FORMAT+x} ]; then
 fi
 
-if [ -z ${DOCKER+x} ] && [ -z ${SONAR+x} ]; then
+if [ -z ${DOCKER+x} ] && [ -z ${SONAR+x} ] &&  [ -z ${ARCH} ]; then
```
This change is not needed because you have an `exit 0` at the end of your block above.

cmbahadir replied:
Fixed with https://github.com/open62541/open62541/pull/3023/commits/5f31676c03d8fe957e07f31f0cd336b460b4cec9

Pro commented on tools/travis/travis_linux_script.sh:
```diff
@@ -1,6 +1,21 @@
+    cp /home/travis/build/open62541/open62541/build/open62541.c components/open62541lib
+    sed -i '/#define UA_IPV6 LWIP_IPV6/d' /home/travis/build/open62541/open62541/build/open62541.h
+    cp /home/travis/build/open62541/open62541/build/open62541.h components/open62541lib/include
+    rm -r components/tcpip_adapter components/lwip components/freertos
```
Can you add some comments why one needs to remove these folders?

cmbahadir replied:
Fixed with https://github.com/open62541/open62541/pull/3023/commits/5f31676c03d8fe957e07f31f0cd336b460b4cec9

Pro commented on tools/travis/travis_linux_script.sh:
```diff
@@ -1,6 +1,21 @@
+    cp /home/travis/build/open62541/open62541/build/open62541.h components/open62541lib/include
+    rm -r components/tcpip_adapter components/lwip components/freertos
+    cp -r $IDF_PATH/components/tcpip_adapter components/ && cp -r $IDF_PATH/components/lwip components/ && cp -r $IDF_PATH/components/freertos components/ 
+    cp components/freertos/include/freertos/* components/freertos/include
```
This is also a bit strange, to copy include files. If this is necessary, than it means the the include directories are not correctly set somewhere. Either in the code itself (`#include ...`) or in the build setup.

cmbahadir commented:
> Thank you for your Pull Request and your work!
> 
> It would be really nice to include a FreeRTOS build into CI. I added some review comments.
> 
> In general, you are somehow mixing the install script and the main build script.
> 
> Can we do it like this?
> 
> 1. `travis_linux_before_install.sh` downloads and installs the `esp-idf` stuff + xtensa toolchain and prepares all the dependencies
> 2. The environment variables should be set in the build matrix, not in the script itself
> 3. `travis_linux_script.sh` just runs cmake and make to create the shared library
> 4. Optional step in a second PR when the previous ones are implemented: use the xtensa toolchain with an example code to build a flashable binary
> 
> @cmbahadir would you have time to update this PR and do some more tests?



> Thank you for your Pull Request and your work!
> 
> It would be really nice to include a FreeRTOS build into CI. I added some review comments.
> 
> In general, you are somehow mixing the install script and the main build script.
> 
> Can we do it like this?
> 
> 1. `travis_linux_before_install.sh` downloads and installs the `esp-idf` stuff + xtensa toolchain and prepares all the dependencies
> 2. The environment variables should be set in the build matrix, not in the script itself
> 3. `travis_linux_script.sh` just runs cmake and make to create the shared library
> 4. Optional step in a second PR when the previous ones are implemented: use the xtensa toolchain with an example code to build a flashable binary
> 
> @cmbahadir would you have time to update this PR and do some more tests?


Hello @Pro ,

Thanks for the review i will try to fix them but there are a few constraints in my mind;

1. I can fix it.
2. IDF_PATH environment variable can be set after clonning esp-idf. So not sure if i can move it to build matrix. (may not be impossible.)
3. I will move all make and cmake to travis_linux_scripts.
4. Actually this repo of me (https://github.com/cmbahadir/opcua-esp32) is an example code to build a flashable binary but can be removed for now to be considered later.

Pro commented:
@cmbahadir thank you for all the updates!
I'll take your current state and try to make it a bit more compact.

cmbahadir commented:
> @cmbahadir thank you for all the updates!
> I'll take your current state and try to make it a bit more compact.

Hi @Pro  yes please i need help. I do not want to occupy the CI more :) and i can fix and squash commits after a review.

Pro commented:
Replaced by #3205
